Jimena ELK blog post (#29)
oktadeveloper/okta-blog: _source/_posts/2019-09-26-getting-started-with-elk.md

_**K**ibana_ provides the visualization front-end, a window into the Elastic Stack. With dashboards and visualization elements, the data stored in Elasticsearch can be explored, aggregated and analyzed.

From version 7 ahead, the ELK Stack was renamed to [**Elastic Stack**](https://www.elastic.co/elk-stack) and added Beats to the stack. Beats is a family of lightweight data shippers that work with Elasticsearch and Logstash. 

## Set up the Elastic Stack

Elastic has published [**a Docker Compose configuration**](https://github.com/elastic/stack-docker), to demonstrate the stack components on a single machine.
Install [Docker](https://docs.docker.com/install/) and [Docker Compose](https://docs.docker.com/compose/install/) and follow these steps to start up the stack:

> Windows users must configure 2 environment variables, check out the instructions on the [**stack-docker** github repository](https://github.com/elastic/stack-docker)

> Allow at least 4GB of RAM for the containers, also check out instructions for your environment

1. Clone the `stack-docker` repository
```
git clone https://github.com/elastic/stack-docker.git
```

> 2. Setup the stack with Docker Compose

Author: good catch! #2 rendered properly, but the ones below it didn't. Fixing....

```

> You may notice exceptions in the log output. For this demonstration, they can be safely ignored. If you run into any issues with docker, you can start fresh with:
```
docker container ls -a | cut -c1-12 | xargs docker container rm --force
docker images | cut -c69-80 | xargs docker rmi
docker system prune -a
```
**NOTE:** This will destroy all docker containers, images and networks, so use at your own risk.

4. Go to [**http://localhost:5601/**](http://localhost:5601/) to log into Kibana.

> Once you log in (using the **elastic** user and the password you captured above), explore the installed dashboards from the Dashboards section via the menu on the left. Heartbeat a Beat that monitors services uptime from a provided list of URLs. Open the dashboard *Heartbeat HTTP monitoring* and see the power of the stack for data visualization.

Author: I'll update to: Heartbeat is one of the Beat services that monitors...

{% img blog/getting-started-with-elk/http-heartbeat.png alt:"Heartbeat HTTP Monitoring" width:"800" %}{: .center-image }

## The JHipster Console

The Jhipster Console, an awesome monitoring solution based on the Elastic Stack, allows the visualization and analysis of JHipster applications metrics over time. The Console provides pre-configured dashboards to monitor microservices infrastructure. You can review the complete list of features at the [**JHipster Console**](https://www.jhipster.tech/monitoring/#jhipster-console) documentation.

> One of the easier ways to start with the JHipster Console is to deploy the applications and enable monitoring with the [**docker-compose sub-generator**](https://www.jhipster.tech/docker-compose/). We'll use this to:

Author: I'll do this one manually because it's another case of bolding a link... sigh...

  price BigDecimal required min(0),
  image ImageBlob
}

relationship ManyToOne {
  Blog{user(login)} to User,
  Post{blog(name)} to Blog
}

relationship ManyToMany {
  Post{tag(name)} to Tag{post}
}

paginate Post, Tag with infinite-scroll
paginate Product with pagination

microservice Product with store
microservice Blog, Post, Tag with blog
```

Now, in your `jhipster` folder, run the [**import-jdl** generator](https://www.jhipster.tech/jdl/#importingjdl).

```SHELL
jhipster import-jdl apps.jh
```

### Deploy Monitoring using `docker-compose`

In the project folder, create a subfolder for the `docker-compose` configuration and run the subgenerator.

```SHELL
mkdir docker-compose
cd docker-compose
jhipster docker-compose
```
The generator will ask you to define the following configurations:

1. Type of application: **Microservices application**
2. Type of gateway: **JHipster based on Zuul**
3. Which applications to include: **blog**, **gateway**, **store**
4. If the database is clustered: **no**
5. if monitoring must be enabled: **yes, with JHipster console**
6. Additional technologies for monitoring: **Zipkin**
7. Password for JHipster Registry: **default**

> [![asciicast](https://asciinema.org/a/261254.svg)](https://asciinema.org/a/261254)

Author: It works!

    @Bean
    @Profile({JHipsterConstants.SPRING_PROFILE_DEVELOPMENT, JHipsterConstants.SPRING_PROFILE_PRODUCTION})
    public LoggingAspect loggingAspect(Environment env) {
        return new LoggingAspect(env);
    }
}
```
[**Zipkin**](https://zipkin.io/) is a distributed tracing system that helps troubleshooting latency issues in microservices architectures. With a traceId propagated from service to service, calls to different services can be correlated and analyzed as a part of the same flow. The Zipkin server and UI are provided with the JHipster Console, and JHipster apps can integrate with Zipkin through [**Spring Cloud Sleuth**](https://cloud.spring.io/spring-cloud-sleuth/reference/html/#sleuth-with-zipkin-via-http). 
To enable Zipkin tracing, add the zipkin profile to `blog-app`, `gateway-app` and `store-app` in **docker-compose/docker-compose.yml**.

```yaml
- SPRING_PROFILES_ACTIVE=prod,swagger,zipkin
```
You also need to rebuild the docker images with the **zipkin** profile, for `blog-app`, `store-app` and `gateway-app` with the following Maven command:
```SHELL
./mvnw package -Pprod -Pzipkin verify jib:dockerBuild -DskipTests
```

> >ProTip:  If you're on a system with bash shell, like Linux or MacOs, you can do this from the `jhipster` folder to build each project at once:

Author: I like having this as a callout rather than just a bolded section

```
for i in blog gateway store
do 
cd $i 
./mvnw package -Pprod -Pzipkin verify jib:dockerBuild -DskipTests
cd ..
done
```

### Run the Monitored Microservices Architecture

Are you ready for the best? Go to the `docker-compose` folder and start the services with the following command:

> ```SHELL

Reviewer: ```suggestion
```shell
```

docker-compose up
```

> You will see a huge amount of logging while each service starts.

Reviewer: ```suggestion
You will see a lot of logging while each service starts.
```

Author: this seems more like a personal style thing. imma leave it.

```
jhipster-registry_1           | ----------------------------------------------------------
jhipster-registry_1           | 	Application 'jhipster-registry' is running! Access URLs:
jhipster-registry_1           | 	Local: 		http://localhost:8761
jhipster-registry_1           | 	External: 	http://172.20.0.2:8761
jhipster-registry_1           | 	Profile(s): 	[composite, dev, swagger, oauth2]
jhipster-registry_1           | ----------------------------------------------------------
```

> Log into the JHipster Registry at [**http://localhost:8761**](http://localhost:8761) with Okta user credentials and check the service's health.

Reviewer: ```suggestion
Log into the JHipster Registry at `http://localhost:8761` with Okta user credentials and check the service's health.
```

{% img blog/getting-started-with-elk/registry.png alt:"JHipster Registry" width:"800" %}{: .center-image }

> Once all the services are up, log into the gateway application and create some blogs and posts to generate traffic. To do this, use the **Entities** menu at the top left of the application. The gateway home is at [**http://localhost:8080**](http://localhost:8080)

Reviewer: ```suggestion
Once all the services are up, log in to the gateway application and create some blogs and posts to generate traffic. To do this, use the **Entities** menu at the top left of the application. The gateway home is at `http://localhost:8080`.
```

> The fun part! Access the JHipster Console at [**http://localhost:5601**](http://localhost:5601). Go to the Dashboards section, and open the **requests-dashboard**. You should see some nice curves:

Reviewer: We should never bold links intentionally. If we wanted bold links, we'd do it in our CSS.
```suggestion
The fun part! Access the JHipster Console at `http://localhost:5601`. Go to the Dashboards section, and open the **requests-dashboard**. You should see some nice curves:
```

{% img blog/getting-started-with-elk/requests-dashboard.png alt:"Requests Dashboard" width:"800" %}{: .center-image }

> Since we integrated the JHipster Console with Zipkin UI, in the traces-dashboard, you find the longest traces duration on the left. If you click on a traceId on the right, it will open the trace in the UI and you will be able to examine the flow.

Reviewer: ```suggestion
Since you integrated the JHipster Console with Zipkin UI, in the traces-dashboard, you can find the longest traces duration on the left. If you click on a traceId on the right, it will open the trace in the UI and you will be able to examine the flow.
```

---
layout: blog_post
title: "Get Started with the ELK Stack"
author: jimena-garbarino

> description: "Learn how to monitor your microservices architecture with the Elastic Stack (Formerly ELK Stack)"

Reviewer: Needs period at the end of the description.
```suggestion
description: "Learn how to monitor your microservices architecture with the Elastic Stack (Formerly ELK Stack)."
```

    {% img blog/getting-started-with-elk/http-heartbeat.png alt:"Heartbeat HTTP Monitoring" width:"800" %}{: .center-image }

## The JHipster Console

The Jhipster Console, an awesome monitoring solution based on the Elastic Stack, allows the visualization and analysis of JHipster applications metrics over time. The Console provides pre-configured dashboards to monitor microservices infrastructure. You can review the complete list of features in [JHipster Console's documentation](https://www.jhipster.tech/monitoring/#jhipster-console).

One of the easier ways to start with the JHipster Console is to deploy the applications and enable monitoring with the [docker-compose sub-generator](https://www.jhipster.tech/docker-compose/). You'll use this to:
- Create a microservices architecture with JHipster
- Enable monitoring with JHipster Console
- Configure OpenID Connect for authentication to microservices

## Create a Java Microservices Architecture with JHipster

> > To install a version of JHipster that will work here, you need to install [**Node.js**](https://nodejs.org/en/)

Reviewer: ```suggestion
> To install a version of JHipster that will work here, you need to install [Node.js](https://nodejs.org/en/).
```

### Install JHipster

```SHELL
npm install -g generator-jhipster@6.3.1
jhipster --version
```

The version command should output something like this:
```
INFO! Using JHipster version installed globally
6.3.1
```

Create a directory for the project:

> ```SHELL

Reviewer: ```suggestion
```shell
```

mkdir jhipster
cd jhipster
```

> Create `apps.jh` to define  store, blog and gateway microservices in JHipster Domain Language (JDL). We are going to recreate [a java-based example of **Microservices Architecture** we have built before](/blog/2019/05/23/java-microservices-spring-cloud-config) for this tutorial.

Reviewer: ```suggestion
Create `apps.jh` to define store, blog, and gateway microservices in JHipster Domain Language (JDL). You are going to recreate [a java-based example of Microservices Architecture we have built before](/blog/2019/05/23/java-microservices-spring-cloud-config) for this tutorial.
```

Once you log in, click **Your Org**, and it will take you to the **Developer Console**. Go to the **Applications** section and add a new **Web Application**. Set the following authentication settings:
- Name: give a name for your application
- Base URIs: `http://localhost:8761` and `http://localhost:8080`
- Login redirect URIs: `http://localhost:8080/login/oauth2/code/oidc` and  `http://localhost:8761/login/oauth2/code/oidc`
- Grant Type Allowed: Authorization Code and Refresh Token

> For simplicity, this tutorial only creates Web App, and its credentials will be used for all the services. In a real environment, each service must identify itself with its own credentials, and you should create one Web App or Service for each one of them in the Okta console.

Copy the **Client ID** and **Client secret**, as we will use it for the application's setup. Find the **Org URL** at the top right corner in the Okta Dashboard.

Create a `docker-compose/.env` file with the following content:

```
OIDC_CLIENT_ID=<client_id>
OIDC_CLIENT_SECRET=<client_secret>
RESOURCE_ISSUER_URI=<org_url>/oauth2/default
```

> Edit `docker-compose/docker-compose.yml` and update the `SECURITY_*` settings for the services `blog-app`, `gateway-app` and `store-app`:

Reviewer: ```suggestion
Edit `docker-compose/docker-compose.yml` and update the `SECURITY_*` settings for the services `blog-app`, `gateway-app`, and `store-app`:
```

```shell
SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI=${RESOURCE_ISSUER_URI}
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_ID=${OIDC_CLIENT_ID}
SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET=${OIDC_CLIENT_SECRET}
```

The same authentication must be set up for the JHipster Registry. Edit `docker-compose/jhipster-registry.yml` and set the same values as the environment section of the `gateway-app`.

JHipster applications require the specific user roles `ROLE_USER` and `ROLE_ADMIN` as claims in the ID Token. In the Okta Developer Console, go to **Users** > **Groups** and create a group for each JHipster role, and add users to each group.

Now go to **API** > **Authorization Servers**, select the **default** server and **Add Claim** with the following settings:
1. Name: groups
2. Include in token type: ID Token, Always
3. Value type: Groups

> 4. Filter: Matches regex, set the Regex `.*`

Reviewer: ```suggestion
4. Filter: Matches regex, set the Regex to `.*`
```

### Enable Debug Logs and Zipkin

> To send debug logs to the JHipster Console, let's update the log level in the prod profile.  Edit **src/main/resources/config/application-prod.yml** to set the level for each service (`blog-app`, `store-app` and `gateway-app`) to  **DEBUG** for the **com.okta.developer.\*** logger. For example, in the blog application-prod.yml:

Reviewer: ```suggestion
To send debug logs to the JHipster Console, let's update the log level in the prod profile.  Edit `src/main/resources/config/application-prod.yml` to set the level for each service (`blog-app`, `store-app`, and `gateway-app`) to  **DEBUG** for the `com.okta.developer.*` logger. For example, in the blog's `application-prod.yml`:
```

```yaml
logging:
    level:
        com.okta.developer.blog: DEBUG
```

> Also, for each service, update the **LoggingAspectConfiguration** to load when the prod profile is active. Change the `@Profile` annotation:

Reviewer: ```suggestion
Also, for each service, update the `LoggingAspectConfiguration` to load when the prod profile is active. Change the `@Profile` annotation:
```

```java
@Configuration
@EnableAspectJAutoProxy
public class LoggingAspectConfiguration {

    @Bean
    @Profile({JHipsterConstants.SPRING_PROFILE_DEVELOPMENT, JHipsterConstants.SPRING_PROFILE_PRODUCTION})
    public LoggingAspect loggingAspect(Environment env) {
        return new LoggingAspect(env);
    }
}
```
[Zipkin](https://zipkin.io/) is a distributed tracing system that helps to troubleshoot latency issues in microservices architectures. With a traceId propagated from service to service, calls to different services can be correlated and analyzed as a part of the same flow. The Zipkin server and UI are provided with the JHipster Console, and JHipster apps can integrate with Zipkin through [Spring Cloud Sleuth](https://cloud.spring.io/spring-cloud-sleuth/reference/html/#sleuth-with-zipkin-via-http). 

> To enable Zipkin tracing, add the `zipkin` profile to `blog-app`, `gateway-app` and `store-app` in `docker-compose/docker-compose.yml`.

Reviewer: ```suggestion
To enable Zipkin tracing, add the `zipkin` profile to `blog-app`, `gateway-app`, and `store-app` in `docker-compose/docker-compose.yml`.
```

```yaml
- SPRING_PROFILES_ACTIVE=prod,swagger,zipkin
```

> You also need to rebuild the Docker images with the `zipkin` profile, for `blog-app`, `store-app` and `gateway-app` with the following Maven command:

Reviewer: ```suggestion
You also need to rebuild the Docker images with the `zipkin` profile, for `blog-app`, `store-app`, and `gateway-app` with the following Maven command:
```

{% img blog/getting-started-with-elk/requests-dashboard.png alt:"Requests Dashboard" width:"800" %}{: .center-image }

Since you integrated the JHipster Console with Zipkin UI, in the traces-dashboard, you can find the longest traces duration on the left. If you click on a traceId on the right, it will open the trace in the UI and you will be able to examine the flow.

{% img blog/getting-started-with-elk/zipkin-ui.png alt:"zipkin ui" width:"800" %}{: .center-image }

## Learn More about JHipster and Elastic Stack

I hope you enjoyed this tutorial and the power of the **Elastic Stack** and the **JHipster Console** for monitoring a microservices architecture.
To continue expanding your knowledge on JHipster monitoring and Okta integration with the Elastic Stack, check out the following links:

- [JHipster Console on Github](https://github.com/jhipster/jhipster-console)
- [JHipster Monitoring Documentation](https://www.jhipster.tech/monitoring/)
- [SAML Authentication and the Elastic Stack](https://www.elastic.co/blog/how-to-enable-saml-authentication-in-kibana-and-elasticsearch)

> - [Authentication in Kibana](https://www.elastic.co/guide/en/kibana/current/kibana-authentication.html)

Reviewer: ```suggestion
- [Authentication in Kibana](https://www.elastic.co/guide/en/kibana/current/kibana-authentication.html)

If you liked this post, chances are you'll like our other posts on microservices and JHipster:

- [Better, Faster, Lighter Java with Java 12 and JHipster 6](/blog/2019/04/04/java-11-java-12-jhipster-oidc)
- [Upgrading Spring Security OAuth and JUnit Tests through the 👀 of a Java Hipster](/blog/2019/04/15/testing-spring-security-oauth-with-junit)
- [Java Microservices with Spring Boot and Spring Cloud](/blog/2019/05/22/java-microservices-spring-boot-spring-cloud)
- [Java Microservices with Spring Cloud Config and JHipster](/blog/2019/05/23/java-microservices-spring-cloud-config)
- [Secure Reactive Microservices with Spring Cloud Gateway](/blog/2019/08/28/reactive-microservices-spring-cloud-gateway)

To be notified when we published new posts, [follow @oktadev on Twitter](https://twitter.com/oktadev). We also publish screencasts to [our YouTube channel](https://www.youtube.com/c/oktadev) on a regular basis.
```
